### PR TITLE
Validate oversampling

### PR DESCRIPTION
**Pull request by timvisee**
Add validation to `oversampling` as described [here](https://github.com/qdrant/qdrant/pull/2070#discussion_r1231008054).

There are more things that can be validated in these structs, such as filters/conditions. But that can be done in a follow-up PR.

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### CODE REVIEW

File: lib/segment/src/types.rs [hunk 1]
```diff
@@ -293,11 +294,12 @@ pub const fn default_quantization_oversampling_value() -> Option<f64> {
...
 
 /// Additional parameters of the search
-#[derive(Debug, Deserialize, Serialize, JsonSchema, Clone, Copy, PartialEq, Default)]
+#[derive(Debug, Deserialize, Serialize, JsonSchema, Validate, Clone, Copy, PartialEq, Default)]
 #[serde(rename_all = "snake_case")]
 pub struct SearchParams {
     /// Params relevant to HNSW index
-    /// /// Size of the beam in a beam-search. Larger the value - more accurate the result, more time required for search.
+    /// Size of the beam in a beam-search. Larger the value - more accurate the result, more time required for search.
+    #[validate(range(min = 4))]
```
Comment (timvisee) [inline lib/segment/src/types.rs:302]: Is this a valid limit? `hnsw_ef >= 4`

This is derived from [here](https://github.com/qdrant/qdrant/blob/d8f8a3666b93b9b7c29ec5a0f57869215d27a84f/lib/segment/src/types.rs#L341-L342).

Comment (generall) [reply lib/segment/src/types.rs:302]: I would say no, let's not over-validate

Comment (generall) [reply lib/segment/src/types.rs:302]: It should work with `hnsw_ef > 0`, who knows what for? Maybe some experiments

**Review by generall — CHANGES_REQUESTED** [review]

File: lib/segment/src/types.rs (hunk 1 above)
Comment (timvisee) [reply lib/segment/src/types.rs:302]: Alright! I'll revert it for now and suggest these changes (with experiment) in a new PR to not delay the other changes further.

**Review by agourlay — APPROVED** [review]

**Review by generall — APPROVED** [review]

**Pull request merged**